### DISCUSSION TITLE
Help on Rules: Calculation with variables and permanent triggering

### DISCUSSION THREAD

**Discussion by schlaubstar** [arendst/Tasmota#15621] [Q&A] [answered]
Hi Community,

I am pretty new to writing rules and thus I need help.

** What Device do I have?** 
An "Athom Power Monitoring Plug" that comes pre-flashed with Tasmota, I assume it uses an BL0937 Chip as energy meter, connected to GPIO4.

**What do I want to accomplish? ** 
- I'd like to keep track of the Total Energy passed through the plug in a variable
- In case active consumption exceeds a threshold I'd like to store the last known "Total Energy" value and a timestamp, when the consumption 'started'
- In case active consumption falls below the threshold I'd like to send a MQTT telegram containing start and end timestamp (ideally I could calculate the duration in seconds) and the amount of energy consumed in that period.

**What's are my problems?** 
- When trying to store the total energy consumed in a variable I get updates in very short periods and not, as I had expected, if the total energy consumption changes.

`Rule1 
ON energy#total Do Var1 %value% ENDON`

if I define it as "one-shot" (Rule1 5) then it is only triggered once, and never again. If I leave it as regular rule, it is triggered within very short periods in time (few milliseconds) without a change to Var1 occurring.
What am I doing wrong here ? 

My second issue: 

`Rule2
ON energy#power>50 DO Backlog0 Var2 %timestamp%; Var3 %Mem1% ENDON`

defined as "one-shot" seems to exactly do what I want: When consumption exceeds 50W it stores the timestamp in Var2 and copies the last known value of energy#total into Var3, fine so far.

Now before submitting the MQTT-Telegram I'd like to do something like:

`Rule3
ON energy#power<5 DO Var3=%Var1%-%Var3% ENDON
ON energy#power<5 DO Publish tele/pump_tasmota_xyz/EnergyConsumedEvent {"EndTime":"%timestamp%", "StartTime":"%Var1%", "PowerConsumed":"%Var3%", "Unit":"kWh"} ENDON` 

How can I substract two variables from each other ? Whatever I do, I fail with it. 
Ideally I'd also calculate the difference in seconds between the timestamps but this is not as important as directly substracting the two different "energy#total" values. 

Thanks for all feedbacks.

**Comment by barbudor** [comment, marked as answer]
For calculating diffrences you may be able to use the [`Sub`](https://tasmota.github.io/docs/Commands/#sub) command
Howver beware that in a rule, all `%VarX%` are replaced by their value before the rule starts to be evaluated. 
Alternatively you can use a firmware compiled with [Expressions](https://tasmota.github.io/docs/Rules/#expressions-in-rules) either by compiling yourself or using on such as `tasmota-fullrules` from https://github.com/tasmota/install/tree/main/firmware/unofficial

You won't be able to compute %timestamp% differences because timestamp is a 32 bit integer (Linux epoch) and variables are stored into single precision floating point which only has 24 bits capacity. You may on the other hand use %time% (number of minutes from midnight)

**Reply by schlaubstar** [reply]
Thank you for the hints and the detailed explanations. 

I managed it to at least achieve what I wanted, however I need to look into the constant triggering.

Just in case someone else wants to do something similar here is the complete ruleset:

`Rule1 
ON energy#total Do Var1 %value% ENDON

Rule2
ON energy#power>50 DO Backlog0 Var3 %timestamp%; Var2 %Var1% ENDON

Rule3 
ON energy#power<10 DO Var4 %Var1% ENDON
ON energy#power<10 DO Sub4 %Var2% ENDON
ON energy#power<10 DO Publish tele/NameOfYourDevice/NameOfMessage {"EndTime":"%timestamp%", "StartTime":"%Var3%", "PowerConsumed":"%Var4%", "Unit":"kWh"} ENDON
`
Activate all rules (Rule# 1) and set Rules2 and 3 to one-shot (Rule# 5)

**Reply by pkkrusty** [reply]
@barbudor Didn't know that about the variables and timestamp. How involved would it be to change the first two `var` or first two `mem` variables to 32-bit integers, specifically so that epoch calculations could be performed? Seems like second-precision would be useful in a number of use-cases, over minute-precision.

Not familiar enough with the base code and C++ to know about the complexity involved in that change.

Or maybe a `#define LONG_VARS` that replaces all var and mem with 32-bit integers, at the cost of code space, to give more flexibility to those that want it.

**Reply by sfromis** [reply]
Replacing single-precision floating point by 32-bit integers would be a lot of code change (not so much code space), and bad loss of flexibility around decimals. If you want 32-bit integer processing in Tasmota, you can do that in Berry on ESP32.

Possibly, doing much in calculations in rules on ESP8266 is overextending the platform to cover stuff much better handled in the backend. While you *can* do a bit of calculations in rules, it is just not a great platform for numerical calculations, as it is really focused on control instead of data processing. You could call Tasmota the "hardware interface", not so much a development platform for polishing off data presentation.

**Reply by schlaubstar** [reply]
Hi everyone,

I know about the shortcomings of the ESP-Modules, however: If you can't process enough locally you will shift even the most basic operations to the cloud (backend) and thus send a lot of unnecessary messages, to process timestamps, initial values etc. completely in the backend. 

I simply did not see the Add/Substract/... functionalities in the first place and think, they are sufficient (for my requirements)

However: Being able to track a timespan at least with (approx.) second resolution on the Edge-device, would be a great feature. In my opinion this could also be done by introducing something like "seconds since midnight" and would not necessarily mean that you'd need the ability to calculate the difference from two timestamps.

**Reply by sfromis** [reply]
Strictly speaking, while add/subtract are only 24-bit floating point, using the [Counter](https://tasmota.github.io/docs/Commands/#counter) command, you can get addition and subtraction of 31-bit numbers, also with current ESP8266 Tasmota. This means that if you use `Time 2` to get the Epoch value (seconds since 1970) included with timestamps, you could use rules to pick up these values and calculate differences between two.

Of course this does nothing for the bad argument of worrying about "unnecessary messages", but you are able to calculate seconds between times with the result kept locally. Of course, the year 2038 issue is coming up, but then you have some years to figure out what to do, if the devices are still alive.

**Reply by pkkrusty** [reply]
Thanks for the good idea of using Counter as a way to deal with epoch time.

**Comment by schlaubstar** [comment]
However: There is one issue not really solved, and I'd kindly ask for help again.

If I add:
`Rule1 
ON energy#total Do Var1 %value% ENDON`

and activate the rule, this is (an extract) of what I get:

`22:18:43.198 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:43.206 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:43.801 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:43.810 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:44.403 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:44.411 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:44.004 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:45.000 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:45.670 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:45.677 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:46.211 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:46.219 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:46.812 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:46.819 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:47.414 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:47.421 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:48.011 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:48.019 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:48.628 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:48.635 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:49.233 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:49.241 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:49.833 RUL: ENERGY#TOTAL performs "Var1 0.0149"
22:18:49.840 MQT: stat/unused01_tasmota_0FF39C/RESULT = {"Var1":"0.0149"}
22:18:50.432 RUL: ENERGY#TOTAL performs "Var1 0.0149"`

I thought, that this would only trigger in case the value of "Energy#Total" changes.
I think it is not good, that the plug is updating the variable ~ twice a second

**Reply by sfromis** [reply]
You got it. That's how it works. Just avoid what you don't like.....

If you save the value to a var, you could then make the trigger depend on the new value being different from the old.
`Rule1 ON energy#total!=%Var1% Do Var1 %value% ENDON`
